File: CSC310-Cpp/_solutions/Module09-Dynamic-Programming/debug03/rod_cutting.cpp
// rod_cutting.cpp
// Rod cutting (CLRS 14.1) with top-down memoization. price[i] is the price
// of a piece of length i (1 <= i <= 10). cutRod returns the maximum revenue
// for a rod of length n and records the first piece of an optimal cut so the
// full decomposition can be printed. The driver also counts recursive calls
// to show how much work memoization saves.
#include <algorithm>
#include <iostream>
#include <vector>

const std::vector<int> price = {0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30};
const int MAX_PIECE = 10;

int cutRodAux(int n, std::vector<int>& memo, std::vector<int>& firstPiece, long& calls) {
    ++calls;
    if (memo.at(n) >= 0) {
        return memo[n];
    }
    int best = 0;
    if (n > 0) {
        best = -1;
        for (int i = 1; i <= std::min(n, MAX_PIECE); ++i) {
            int candidate = price[i] + cutRodAux(n - i, memo, firstPiece, calls);
            if (candidate > best) {
                best = candidate;
                firstPiece[n] = i;
            }
        }
    }
    memo[n] = best;
    return best;
}

int cutRod(int n, std::vector<int>& firstPiece, long& calls) {
    std::vector<int> memo(n + 1, -1);
    firstPiece.assign(n + 1, 0);
    calls = 0;
    return cutRodAux(n, memo, firstPiece, calls);
}

int main() {
    for (int n : {1, 2, 3, 4, 7, 10, 13}) {
        std::vector<int> firstPiece;
        long calls = 0;
        int revenue = cutRod(n, firstPiece, calls);
        std::cout << "length " << n << ": revenue " << revenue << ", pieces:";
        int rest = n;
        while (rest > 0 && firstPiece[rest] > 0) {
            std::cout << ' ' << firstPiece[rest];
            rest -= firstPiece[rest];
        }
        std::cout << "   (" << calls << " calls)\n";
    }
    return 0;
}

File: CSC310-Python/Module09-Dynamic-Programming/debug03/rod_cutting.py
"""Rod cutting (CLRS Section 14.1).

memoized_cut_rod is the top-down algorithm MEMOIZED-CUT-ROD; it records
each subproblem's answer in r[] so that no rod length is solved twice.
extended_bottom_up_cut_rod is the bottom-up version that also records
the first piece of an optimal cut so the cuts can be printed.  A call
counter shows how much work the top-down version does.
"""

import time

calls = 0


def memoized_cut_rod(p, n):
    r = [float("-inf")] * (n + 1)
    return _cut_rod_aux(p, n, r)


def _cut_rod_aux(p, n, r):
    global calls
    calls += 1
    if n == 0:
        q = 0
    else:
        q = float("-inf")
        for i in range(1, n + 1):
            q = max(q, p[i] + _cut_rod_aux(p, n - i, r))
    if r[n] >= 0:
        return r[n]
    r[n] = q
    return q


def extended_bottom_up_cut_rod(p, n):
    r = [0] * (n + 1)
    s = [0] * (n + 1)
    for j in range(1, n + 1):
        q = float("-inf")
        for i in range(1, j + 1):
            if q < p[i] + r[j - i]:
                q = p[i] + r[j - i]
                s[j] = i
        r[j] = q
    return r, s


def cuts(s, n):
    pieces = []
    while n > 0:
        pieces.append(s[n])
        n -= s[n]
    return pieces


def main():
    global calls
    # CLRS Section 14.1 price table for lengths 1..10, extended linearly beyond.
    p = [0, 1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
    p += [30 + 3 * (i - 10) for i in range(11, 31)]

    r, s = extended_bottom_up_cut_rod(p, 10)
    print(" n  revenue  calls  cuts")
    for n in range(1, 11):
        calls = 0
        rev = memoized_cut_rod(p, n)
        print(f"{n:>2}  {rev:>7}  {calls:>5}  {cuts(s, n)}")

    calls = 0
    start = time.perf_counter()
    rev = memoized_cut_rod(p, 30)
    elapsed = time.perf_counter() - start
    print()
    print(f"rod of length 30: revenue {rev}, {calls} calls, {elapsed:.3f} s")


if __name__ == "__main__":
    main()
